File: MeanFieldDyn.m
%%
clear

global deltaQ;
global deltaT;
global tau;
global eta;

deltaQ = 1e-2;
deltaT = 1e-2;
%% Set up game

tau = 10;
eta = 1;

game = zeros(2, 2, 3);
game(:, :, 1) = [3, 0; 0, 1];
game(:, :, 2) = [3, 5; 5, 1];
nAgents = 1e3;
qValues = rand(2, nAgents);

%% Determine p_init

p = @(Q)  nnz(find(((qValues(1, :) > Q(1)) & (qValues(1, :) < (Q(1) + deltaQ)) & (qValues(2, :) > Q(2)) & (qValues(2, :) < (Q(2) + deltaQ)))))/nAgents;

%% Determine v(t)

v = @(Q, X) eta/(exp(tau * (Q(1))) + exp(tau * (Q(2)))) * [exp(tau * (Q(1))); exp(tau * (Q(2)))] .* (game(:, :, 2) * X - Q);

%% Determine p_dot

pdot = @(Q, X) -1 * sum((p(Q + deltaQ) * v(Q + deltaQ, X) - p(Q) * v(Q, X))./deltaQ);

%% Update p(t)

for t = 0:deltaT:10
   x = [0; 0];            
    for q1 = 0:deltaQ:(1 - deltaQ)
        for q2 = 0:deltaQ:(1 - deltaQ)
            x(1) = x(1) + ((exp(tau * (q1 + deltaQ/2))/(exp(tau * (q1 + deltaQ/2)) + exp(tau * (q2 + deltaQ/2)))) * p([q1; q2]));
            x(2) = x(2) + ((exp(tau * (q2 + deltaQ/2))/(exp(tau * (q1 + deltaQ/2)) + exp(tau * (q2 + deltaQ/2)))) * p([q1; q2]));
        end
    end
   
   xCurrent = x
    
   p_old = p;   
   p = @(Q) p_old(Q) + pdot(Q, xCurrent) * deltaT;
   
   v = @(Q, X) eta/(exp(tau * (Q(1))) + exp(tau * (Q(2)))) * [exp(tau * (Q(1))); exp(tau * (Q(2)))] .* (game(:, :, 1) * X - Q);
   pdot = @(Q, X) -1 * sum((p_old(Q + deltaQ) * v(Q + deltaQ, X) - p_old(Q) * v(Q, X))./deltaQ);
end

File: EGT-RL/MeanFieldDyn.py
import numpy as np
from pClass import p
from tqdm import tqdm

## Set up code parameters
tau = 10

## Set up game
game = np.zeros((2, 2, 2))
game[:, :, 0] = np.array([[3, 0], [5, 1]])
game[:, :, 1] = np.array([[3, 5], [0, 1]])

nAgents = 1e3
qValues = np.random.rand(2, int(nAgents))
meanStrategy = [np.mean([np.exp(tau * qValues[0, i])/(np.exp(tau * qValues[0, i]) + np.exp(tau * qValues[1, i])) for i in range(int(nAgents))]), np.mean([np.exp(tau * qValues[1, i])/(np.exp(tau * qValues[0, i]) + np.exp(tau * qValues[1, i])) for i in range(int(nAgents))])]

## Generate pInit

P = p(game, meanStrategy=meanStrategy, initP=True, initQ=qValues)

for t in tqdm(np.arange(0, 0.5, P.deltaT)):

    meanStrategy = np.zeros(2)
    for q1 in np.arange(0, 1, P.deltaQ):
        for q2 in np.arange(0, 1, P.deltaQ):
            meanStrategy[0] += np.exp(P.tau * (q1 + P.deltaQ/2))/(np.exp(P.tau * (q1 + P.deltaQ/2)) + np.exp(P.tau * (q2 + P.deltaQ/2))) * P([q1, q2])

            meanStrategy[1] += np.exp(P.tau * (q2 + P.deltaQ/2))/(np.exp(P.tau * (q1 + P.deltaQ/2)) + np.exp(P.tau * (q2 + P.deltaQ/2))) * P([q1, q2])
    print(sum(meanStrategy))

    P = p(game, oldP=P, meanStrategy=meanStrategy)